internal/s3/controller.go

```diff
@@ -0,0 +1,250 @@
+package s3
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/labstack/echo/v4"
+
+	"github.com/packethost/aws-s3-proxy/internal/config"
+)
+
+var getStatus *regexp.Regexp
+
+func init() {
+	getStatus = regexp.MustCompile(`status code: (?P<status>\d\d\d),`)
+}
+
+func storeObject(e echo.Context, r io.Reader, path *string) error {
+	c := config.Cfg
+
+	// tee the stream
+	var buf bytes.Buffer
+	tee := io.TeeReader(r, &buf)
+
+	err := e.Stream(http.StatusOK, echo.MIMEOctetStream, tee)
+	if err != nil {
+		e.Error(err)
+
+		return err
+	}
+
+	c.Logger.Debugf("writing object %s to local cache", *path)
+
+	_, err = put(context.Background(), &c.PrimaryStore, path, bytes.NewReader(buf.Bytes()))
+	if err != nil {
+		c.Logger.Error("read through cache save failed")
+
+		return err
+	}
+
+	return nil
+}
+
+func trySecondary(e echo.Context) error {
+	c := config.Cfg
+	h := c.HTTPOpts
+	req := e.Request()
+	res := e.Response()
+	path := &req.URL.Path
+
+	// Range header
+	var rangeHeader *string
+	if candidate := req.Header.Get("Range"); candidate != "" {
+		rangeHeader = &candidate
+	}
+
+	get, err := get(req.Context(), &c.SecondaryStore, path, rangeHeader)
+	if err != nil {
+		status := http.StatusInternalServerError
+
+		if getStatus.Match([]byte(err.Error())) {
+			statusStr := getStatus.FindStringSubmatch(err.Error())
+
+			status, _ = strconv.Atoi(statusStr[1])
+		}
+
+		return e.String(status, err.Error())
+	}
+
+	// stream object to client
+	setHeadersFromAwsResponse(res, get, h.HTTPCacheControl, h.HTTPExpires)
+
+	if c.ReadThrough.CacheToPrimary {
+		return storeObject(e, get.Output.Body, path)
+	}
+
+	return e.Stream(http.StatusOK, echo.MIMEOctetStream, get.Output.Body)
+}
+
+// AwsS3Get handles download requests
+func AwsS3Get(e echo.Context) error {
+	c := config.Cfg
+	h := c.HTTPOpts
+	req := e.Request()
+	res := e.Response()
+	path := &req.URL.Path
+
+	// Range header
+	var rangeHeader *string
+	if candidate := req.Header.Get("Range"); candidate != "" {
+		rangeHeader = &candidate
+	}
+
+	get, err := get(req.Context(), &c.PrimaryStore, path, rangeHeader)
+	if err != nil {
+		if c.ReadThrough.Enabled {
+			e.Logger().Error("err in primary, trying secondary")
+			return trySecondary(e)
```

Comment:
Should we blindly always try secondary?

Also, should we log the error of the first attempt? That might get a little noisy if you're expecting the secondary to serve the request. But perhaps, only log the error unless the error was for a missing object?

Reply:
Changing to a warning doesn't really address anything. I was more wondering if we should actually log `err` if `err != NotFound` so that we can become aware of issues unrelated to files not found. Ex: maybe some other error like an auth error is returned. `err in primary, trying secondary` without logging what the error is won't be helpful attempting to troubleshooting issues with the primary.

Reply:
This is desired behavior. If the primary bucket isn't working just try the other. If we want to go and increase logging that can be addressed in a different PR, IMO. 

This PR tho also refactors errors to be better forwarded from Buckets. So an upstream 404 is no longer a 500. This will have downstream benefits as the primary issue consumers have has is opaque `500`'s

Reply:
not disagreeing with always trying the secondary. Just that we should have better logging. but 👍